### PR TITLE
[WIP] Ecto batch fun

### LINKED ISSUE #5: [Ecto] Support for associations which uses through
> https://hexdocs.pm/ecto/Ecto.Schema.html#has_many/3-has_many-has_one-through

Right now a RuntimeError saying the association cannot be found in the schema is raised.

### PR DESCRIPTION
**Pull request by benwilson512**
Fixes #5 

This PR adds the ability to configure customizable batch loading functions to the Ecto data source. The `query` option already made it possible to add scoping conditions when looking up data. What was lacking however was the ability to accumulate values, and then supply a function that could use those values to run arbitrary SQL to do the lookup.

This PR also adds support for `through` associations.

Remaining: 

- [x] basic batch function
- [x] cardinality handling
- [x] through handling.
- [ ] docs

### CODE REVIEW

File: lib/dataloader/ecto.ex [hunk 1]
```diff
@@ -260,31 +317,64 @@ if Code.ensure_loaded?(Ecto) do
         """
       end
 
+      defp normalize_value(queryable, [{col, value}]) do
+        case queryable.__schema__(:primary_key) do
+          [^col] ->
+            {:primary, col, value}
+
+          _ ->
+            {:not_primary, col, value}
```
Comment (benwilson512) [inline lib/dataloader/ecto.ex]: A+ naming.

File: lib/dataloader/ecto.ex [hunk 2]
```diff
@@ -304,6 +397,20 @@ if Code.ensure_loaded?(Ecto) do
...
+        fn value ->
+          value
+        end
+      end
+
+      defp cardinality_mapper(:one) do
+        fn
+          [] -> nil
+          [value] -> value
+          [_ | _] -> raise "expected one but got many"
```
Comment (benwilson512) [inline lib/dataloader/ecto.ex]: Should maybe raise the same error as Repo.one

File: test/dataloader/ecto_test.exs [hunk 3]
```diff
@@ -195,8 +238,41 @@ defmodule Dataloader.EctoTest do
     assert_receive(:querying)
   end
 
-  defp query(queryable, _args, test_pid) do
-    send(test_pid, :querying)
-    queryable
+  test "we handle a variety of key possibilities", %{loader: loader} do
```
Comment (benwilson512) [inline test/dataloader/ecto_test.exs]: Handy catalog of the various possibilities.

**Comment by tlvenn** [comment]
@benwilson512 any chance to land this PR on master soon ?

**Comment by benwilson512** [comment]
Wanted to test it for a bit locally, been happy with it in prod, so off to master it goes!

Thanks for the input everyone!

**Comment by axelson** [comment]
@benwilson512 I'm glad this is merged! I've been playing with it for a PoC and it's working well. What are the plans for the docs? If you could help write a skeleton I'd be willing to help flesh it out.

**Pull request merged**